fulltext/extract.py

```diff
@@ -1,57 +1,252 @@
-        metrics.report('ProcessingDuration', duration, units='Microseconds')
-        logger.debug('Finished processing in %i microseconds', duration)
 
+def get_task(identifier: str, id_type: str, version: str) -> Extraction:
+    """
+    Get the status of an extraction task.
+
+    Parameters
+    ----------
+    identifier : str
+        Unique identifier for the paper being extracted. Usually an arXiv ID.
+    id_type : str
+        Either 'arxiv' or 'submission'.
+    version : str
+        Extractor version.
+
+    Returns
+    -------
+    :class:`Extraction`
+
+    """
+    _task_id = task_id(identifier, id_type, version)
+    result = extract.AsyncResult(_task_id)
+    exception: Optional[str] = None
+    owner: Optional[str] = None
+    if result.status == 'PENDING':
+        raise NoSuchTask('No such task')
+    elif result.status in ['SENT', 'STARTED', 'RETRY']:
+        _status = Extraction.Status.IN_PROGRESS
+    elif result.status == 'FAILURE':
+        _status = Extraction.Status.FAILED
+        exception = str(result.result)
+    elif result.status == 'SUCCESS':
+        _status = Extraction.Status.SUCCEEDED
+        owner = str(result.result['owner'])
+    else:
+        raise RuntimeError('Unexpected state')
+    return Extraction(
+        identifier=identifier,
+        bucket=id_type,
+        task_id=_task_id,
+        version=version,
+        status=_status,
+        exception=exception,
+        owner=owner
+    )
+
+
+@celery_app.task
+def extract(identifier: str, id_type: str, version: str,
+            owner: Optional[str] = None,
+            token: Optional[str] = None) -> Dict[str, str]:
+    """Perform text extraction for a single arXiv document."""
+    logger.debug('Perform extraction for %s in bucket %s with version %s',
+                 identifier, id_type, version)
+
+    canonical = pdf.CanonicalPDF.current_session()
+    storage = store.Storage.current_session()
+    compilations = compiler.Compiler.current_session()
+
+    try:
+        if id_type == SupportedBuckets.ARXIV:
+            pdf_path = canonical.retrieve(identifier)
+        elif id_type == SupportedBuckets.SUBMISSION:
+            pdf_path, owner = compilations.retrieve(identifier, token)
+        else:
+            RuntimeError('Unsupported identifier')
+        extraction = storage.retrieve(identifier, version, bucket=id_type,
+                                      meta_only=True)
+        content = do_extraction(pdf_path)
+        assert content is not None
     except Exception as e:
-        logger.error('Failed to process %s: %s' % (document_id, e))
+        logger.error('Failed to process %s: %s', identifier, e)
+        storage.store(extraction.copy(status=Extraction.Status.FAILED,
+                                      ended=datetime.now(UTC),
+                                      exception=str(e)))
         raise e
-    return {
-        'document_id': document_id,
-    }
-
+    finally:
+        os.remove(pdf_path)    # Cleanup.
```

Comment:
This makes me wonder if this is intended to be thread safe. If this is not thread safe put some info about that somewhere.  Sorry if I missed the info about thread safety already in the docs.

Reply:
Can you say more about what makes you wonder about thread safety here? Since this is a Celery task, the assumption is that this function will be executed by an isolated process running a single thread. [Celery uses multiprocessing for concurrency by default](http://docs.celeryproject.org/en/latest/userguide/workers.html#concurrency).

Reply:
The os.remove(pdf_path) makes me think of a race condition between two single threads on the same server processing the same pdf_path.